File: app/capture/webcam_capture.py
import logging
import time

import cv2

_SKIP_FRAMES = 2


class WebcamCapture:
    def __init__(self, logger: logging.Logger):
        self.cap = cv2.VideoCapture(0)
        self.logger = logger

        if not self.cap.isOpened():
            raise OSError("Failed to initialise webcam")
        time.sleep(2)  # wait for webcam to initialise and adjust light levels

    def get_frame(self) -> cv2.Mat | None:
        # skip some frames to reduce frame rate
        for _ in range(_SKIP_FRAMES):
            self.cap.read()
        ret, frame = self.cap.read()
        if not ret:
            self.logger.info("Failed to capture webcam")
            return None
        return frame

    def release(self) -> None:
        self.cap.release()

File: src/capture/webcam_capture.rs
use std::thread::sleep;
use std::time::Duration;
use anyhow::Result;
use opencv::{videoio, core, highgui, prelude::*};


pub struct WebcamCapture {
    pub capture: videoio::VideoCapture,
    pub window_name: String,
}

impl WebcamCapture {
    pub fn new(window_name: &str) -> Result<Self> {
        highgui::named_window(window_name, highgui::WINDOW_AUTOSIZE)?;
        let cam = videoio::VideoCapture::new(0, videoio::CAP_ANY)?;
        let opened = videoio::VideoCapture::is_opened(&cam)?;
        if !opened {
            panic!("Unable to open default camera!");
        }

        highgui::named_window(window_name, highgui::WINDOW_AUTOSIZE)?;
        sleep(Duration::from_millis(1000));
        Ok(Self {
            capture: cam,
            window_name: window_name.to_string(),
        })
    }

    pub fn get_frame(&mut self) -> Result<core::Mat> {
        let mut frame = core::Mat::default();
        self.capture.read(&mut frame)?;
        Ok(frame)
    }

    pub fn release(&mut self) -> Result<()> {
        self.capture.release()?;
        highgui::destroy_window(&self.window_name)?;
        Ok(())
    }
}
